Implementation and tests for conversion to direction cosine matrices

adbugger opened the pull request:
It took a while for me to get the shape and axes for concatenation right. Now, `rotation.as_dcm[i]` represents the matrix form of the `ith` index quaternion in the input. Test  cases also check for single quaternion vs a stack containing a single quaternion.

adbugger committed:
ENH: Implemented as_dcm method.

adbugger committed:
TST: Added test cases for as_dcm

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -84,3 +85,41 @@ def as_quaternion(self):
+    def as_dcm(self):
+        """Return the direction cosine matrix representation of the Rotation.
+
+        This function returns a numpy.ndarray of shape (3 x 3) or (N x 3 x 3)
```
Just something what I thought: maybe write shape as `(3, 3)` and `(N, 3, 3)` just to keep it uniform everywhere in the docstrings. Although this notation also looks nice, I don't mind to keep it here. Judge yourself.

adbugger replied:
I think I'll prefer uniformity over beauty in this case.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -84,3 +85,41 @@ def as_quaternion(self):
+        This function returns a numpy.ndarray of shape (3 x 3) or (N x 3 x 3)
+        depending on the input that was used to initialize the object.
+        """
+        quat = self._quat.copy()
```
Any point to copy `self._quat` here? I mean that just reference assignment should do `quat = self._quat`.

adbugger replied:
Sorry, there's no point. I kept tweaking my implementation to get the shape right, and this was from a previous attempt. I'll remove it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -84,3 +85,41 @@ def as_quaternion(self):
+        yz = y * z
+        xw = x * w
+
+        d1 = np.concatenate(
```
OK, I suggest a simpler approach. I think you will like it more.

1. Create `dcm` as `np.empty` with an appropriate shape.
2. Assign `dcm[:, 0, 0]`, `dcm[:, 0, 1]`, etc. This way you also don't need to do any broadcasting for quaternion components.

I think it will make the conversion procedure easier to understand (as implementation reference) and it will be slightly more efficient as well (including memory allocation).

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -62,3 +62,81 @@ def test_zero_norms_from_quaternion():
+def test_as_dcm_single_2d_quaternion():
+    quat = [[0, 0, 1, 1]]
+    mat = Rotation.from_quaternion(quat).as_dcm()
+    assert(mat.shape == (1, 3, 3))
```
I don't know exact reasons 😄, but it seems to me that `assert_` from `numpy.testing` should be used instead of Python built-in assert. At least this is how it is done, I think it's better to follow.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -84,3 +85,41 @@ def as_quaternion(self):
+        z = quat[:, 2].reshape(-1, 1)
+        w = quat[:, 3].reshape(-1, 1)
+
+        x2 = np.power(x, 2)
```
Really preferential, but `x ** 2` or `x * x` will also do.

nmayorov reviewed (commented):
Looks really good overall. I suggested a slightly easier approach to fill `dcm` array.

I like the test coverage.

adbugger committed:
MAINT: Simpler and more efficient dcm component assignment

According to the tests, array indices during assignment are now
required to be transpose of the earlier intuition.

adbugger committed:
STY: Changed description of shape in docstrings

adbugger committed:
TST: Modify tests to use numpy.assert_equal instead of python assert

nmayorov commented:
Let's wait for checks to finish and then you merge it.
